### PR TITLE
hibernate-create-movie-session-hw HW solution

### PR DESCRIPTION
**Pull request by luk14n**

### CODE REVIEW

File: src/main/java/mate/academy/dao/impl/CinemaHallDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Override
+    public List<CinemaHall> getAll() {
+        try (SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
+                Session session = sessionFactory.openSession()) {
+            Query<CinemaHall> getAllCinemaHallsQuery = session
+                    .createQuery("from CinemaHall", CinemaHall.class);
+            List<CinemaHall> resultList = getAllCinemaHallsQuery.getResultList();
+            resultList.sort(Comparator.comparing(CinemaHall::getCapacity));
```
Comment (Rommelua) [inline src/main/java/mate/academy/dao/impl/CinemaHallDaoImpl.java:58]: we should perform sorting on DB layer if we need it - not in java code

File: src/main/java/mate/academy/dao/impl/MovieDaoImpl.java [hunk 2]
```diff
@@ -45,6 +49,14 @@ public Optional<Movie> get(Long id) {
 
     @Override
     public List<Movie> getAll() {
-        return null;
+        SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
+        try (Session session = sessionFactory.openSession()) {
+            Query<Movie> getAllMoviesQuery = session.createQuery("from Movie", Movie.class);
+            List<Movie> resultList = getAllMoviesQuery.getResultList();
+            resultList.sort(Comparator.comparing(Movie::getTitle));
```
Comment (Rommelua) [inline src/main/java/mate/academy/dao/impl/MovieDaoImpl.java:56]: we should perform sorting on DB layer if we need it - not in java code

File: src/main/java/mate/academy/dao/impl/CinemaHallDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            return Optional.ofNullable(session.get(CinemaHall.class, id));
+        } catch (Exception e) {
+            throw new DataProcessingException("Can't get CinemaHall by ID: " + id, e);
+        }
+    }
+
+    @Override
+    public List<CinemaHall> getAll() {
+        try (SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
+                Session session = sessionFactory.openSession()) {
```
Comment (Rommelua) [inline src/main/java/mate/academy/dao/impl/CinemaHallDaoImpl.java:54]:
```suggestion
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
```

File: src/main/java/mate/academy/dao/impl/MovieDaoImpl.java [hunk 4]
```diff
@@ -45,6 +49,14 @@ public Optional<Movie> get(Long id) {
 
     @Override
     public List<Movie> getAll() {
-        return null;
+        SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
+        try (Session session = sessionFactory.openSession()) {
```
Comment (Rommelua) [inline src/main/java/mate/academy/dao/impl/MovieDaoImpl.java:53]:
```suggestion
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
```

File: src/main/java/mate/academy/dao/impl/MovieSessionDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            throw new DataProcessingException("Can't get MovieSession by ID: " + id, e);
+        }
+    }
+
+    @Override
+    public List<MovieSession> findAvailableSessions(Long movieId, LocalDate date) {
+        try (SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
+                Session session = sessionFactory.openSession()) {
+            LocalDateTime startOfDay = date.atStartOfDay();
+            LocalDateTime endOfDay = date.atTime(23, 59, 59);
```
Comment (Rommelua) [inline src/main/java/mate/academy/dao/impl/MovieSessionDaoImpl.java:64]: use LocalTime.MAX

File: src/main/java/mate/academy/dao/impl/MovieSessionDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            Query<MovieSession> findAvailableSessionsQuery = session.createQuery(
+                    "from MovieSession m "
+                            + "where m.showTime >= :startOfDay "
+                            + "and m.showTime <= :endOfDay "
```
Comment (Rommelua) [inline src/main/java/mate/academy/dao/impl/MovieSessionDaoImpl.java:68]: use BETWEEN

File: src/main/java/mate/academy/model/CinemaHall.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        }
+        CinemaHall that = (CinemaHall) o;
+        return capacity == that.capacity
+                && Objects.equals(id, that.id)
+                && Objects.equals(description, that.description);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, capacity, description);
```
Comment (Rommelua) [inline src/main/java/mate/academy/model/CinemaHall.java:60]: delete these methods

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/dao/impl/CinemaHallDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                transaction.rollback();
+            }
+            throw new RuntimeException("Can't add cinemaHall " + cinemaHall, e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
+        return cinemaHall;
+    }
```
Comment (sokimaaa) [inline src/main/java/mate/academy/dao/impl/CinemaHallDaoImpl.java:37]:
```suggestion
    public CinemaHall add(CinemaHall cinemaHall) {
        Session session = null;
        Transaction transaction = null;
        try {
            session = HibernateUtil.getSessionFactory().openSession();
            transaction = session.beginTransaction();
            session.persist(cinemaHall);
            transaction.commit();
            return cinemaHall;
        } catch (Exception e) {
            if (transaction != null) {
                transaction.rollback();
            }
            throw new RuntimeException("Can't add cinemaHall " + cinemaHall, e);
        } finally {
            if (session != null) {
                session.close();
            }
        }
    }
```

File: src/main/java/mate/academy/model/MovieSession.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+@Entity
+@Table(name = "movie_sessions")
+public class MovieSession {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne (fetch = FetchType.LAZY)
+    private Movie movie;
+    @OneToOne (fetch = FetchType.LAZY)
+    private CinemaHall cinemaHall;
```
Comment (sokimaaa) [inline src/main/java/mate/academy/model/MovieSession.java:21]: Wrong relations between tables
```suggestion
    @OneToOne (fetch = FetchType.LAZY)
    private Movie movie;
    @OneToOne (fetch = FetchType.LAZY)
    private CinemaHall cinemaHall;
```

File: src/main/java/mate/academy/service/impl/CinemaHallServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    private CinemaHallDao cinemaHallDao;
+
+    @Override
+    public CinemaHall add(CinemaHall cinemaHall) {
+        return cinemaHallDao.add(cinemaHall);
+    }
+
+    @Override
+    public CinemaHall get(Long id) {
+        return cinemaHallDao.get(id).get();
```
Comment (sokimaaa) [inline src/main/java/mate/academy/service/impl/CinemaHallServiceImpl.java:22]: Never use Optional#get without the previous null-check. 
Consider reading why do we need Optional. 
```suggestion
        return cinemaHallDao.get(id).get();
```

**Review by sokimaaa — CHANGES_REQUESTED** [review]

**Review by sokimaaa — APPROVED** [review]

**Pull request closed without merging**